solr/core/src/java/org/apache/solr/update/processor/UpsertCondition.java

```diff
@@ -0,0 +1,276 @@
+        case MUST:
+          if (!ruleMatched) {
+            return false;
+          }
+          atLeastOneMatched = true;
+          break;
+        case MUST_NOT:
+          if (ruleMatched) {
+            return false;
+          }
+          atLeastOneMatched = true;
+          break;
+        default:
+          atLeastOneMatched = ruleMatched || atLeastOneMatched;
+          break;
+      }
+    }
+    return atLeastOneMatched;
+  }
+
+  enum Action {
+    UPSERT,
+    INSERT,
+    SKIP;
+  }
+
+  private static class Docs {
+    private final SolrInputDocument oldDoc;
+    private final SolrInputDocument newDoc;
+
+    Docs(SolrInputDocument oldDoc, SolrInputDocument newDoc) {
+      this.oldDoc = oldDoc;
+      this.newDoc = newDoc;
+    }
+
+    SolrInputDocument getOldDoc() {
+      return oldDoc;
+    }
+
+    SolrInputDocument getNewDoc() {
+      return newDoc;
+    }
+  }
+
+  private static class FieldRule {
+    private static final Pattern RULE_CONDITION_PATTERN = Pattern.compile("^(OLD|NEW)\\.(\\*|\\w+:(?:\\w+|\\*))$");
+
+    private final BooleanClause.Occur occur;
+    private final Function<Docs, SolrInputDocument> docGetter;
+    private final Predicate<SolrInputDocument> docPredicate;
+
+    private FieldRule(BooleanClause.Occur occur, Function<Docs, SolrInputDocument> docGetter, Predicate<SolrInputDocument> docPredicate) {
+      this.occur = occur;
+      this.docGetter = docGetter;
+      this.docPredicate = docPredicate;
+    }
+
+    static FieldRule parse(BooleanClause.Occur occur, String condition) {
+      Matcher m = RULE_CONDITION_PATTERN.matcher(condition);
+      if (m.matches()) {
+        String doc = m.group(1);
+        String predicate = m.group(2);
+        Function<Docs, SolrInputDocument> docGetter;
+        if (doc.equalsIgnoreCase("OLD")) {
+          docGetter = Docs::getOldDoc;
+        } else {
+          docGetter = Docs::getNewDoc;
+        }
+
+        Predicate<SolrInputDocument> docPredicate;
+        if ("*".equals(predicate)) {
+          docPredicate = Objects::nonNull;
+        } else {
+          String[] parts = predicate.split(":");
+          String field = parts[0];
+          String value = parts[1];
+          if ("*".equals(value)) {
+            docPredicate = forField(field, Objects::nonNull);
+          } else {
+            docPredicate = forField(field, value::equals);
+          }
+        }
+        return new FieldRule(occur, docGetter, docPredicate);
+      }
+      throw new SolrException(SERVER_ERROR, "'" + condition + "' not a valid condition for rule");
+    }
+
+    BooleanClause.Occur getOccur() {
+      return occur;
+    }
+
+    boolean matches(Docs docs) {
+      SolrInputDocument doc = docGetter.apply(docs);
+      return docPredicate.test(doc);
+    }
+
+    private static Predicate<SolrInputDocument> forField(String field, Predicate<Object> fieldPredicate) {
+      return doc -> doc != null && fieldPredicate.test(doc.getFieldValue(field));
```

Comment:
Not sure how this will work for non-String fields and multi-valued fields. `getFieldValue` returns an Object which could be an Integer etc. You might have to toString it first? And maybe also use `getFieldValues` for multi-valued fields.

Reply:
I was only thinking about single valued string fields, but I should probably make it handle multi-valued and non-string fields too.  There's no reason not to really.